src/blog/first-100k-lines.vmd

```diff
@@ -0,0 +1,415 @@
+
+
+If you don't have a vast suite of tests, your project probably won't survive its first 100,000 lines. [# There are exceptions to this. In game development, requirements change so much, that even end-to-end tests can have a negative return-on-investment. Other methods of testing are preferred, such as setting two random AI players against each other. Be sure to save the random seed!]
+
+
+Some languages are easier to test with. Javascript's [Monkey Patching](https://www.audero.it/blog/2016/12/05/monkey-patching-javascript/) is a wonderful alternative to mocking, and can make testing much easier.
+
+
+## Prefer End-to-End Tests
+
+In the early days of the Vale compiler, we had a lot of unit tests for our various components.
+
+For those unfamiliar, a *unit test* is one that specifically tests just one piece of code. You craft some inputs, feed those into your code, and check the outputs.
+
+Unit tests are nice because they tell you _exactly_ where the bug is, because they test only a small piece of your code.
+
+
+However, the data being passed between these components was changing very often, because the project was evolving rapidly in response to user feedback and experiments.
+
+*And unfortunately, every time this happened, we had to change the unit tests.* How irksome!
+
+
+Instead, we've switched over to *end-to-end tests.* An end-to-end test is where a script will open up your application and click on buttons and type inputs in the right sequence to indirectly run some specific code in your program.
+
+For the Vale compiler, it means we run the compiler with some Vale source code, then run it, and make sure it produces the right output. As of this writing, the Vale compiler has *1,033 end-to-end tests.*
+
+
+Some caveats on this advice:
+
+ * It not apply to larger endeavors, only ones that have rapidly shifting internals, as new projects often do. As a project becomes larger, other priorities dominate.
+ * If there's a piece of the program whose code changes often but the inputs and outputs are relatively stable, unit tests are still a good choice.
+
+
+
+# Prioritize Development Velocity
+
+If you're not careful, your development speed can slow to a crawl. Here are some ways to keep yourself nimble:
+
+
+*Use a language with good compile speeds.* If you go doomscrolling on reddit while you're building, you know your compile times are too long.
+
+
+*Write more tests!* The sooner you can find a bug, the easier it will be to fix.
+
+
+*Use a memory-safe language.* Memory safety doesn't just help with security, it helps you avoid bugs that are very difficult to diagnose.
+
+
+*Use a flexible language.* I know it's taboo to say this, but be wary of using functional programming languages or Rust, unless you have many years of experience architecting with them and know how to avoid the API churn and refactor shockwaves induced by the added constraints.
```

Comment:
I don't really recommend criticizing people's language choices in an off-hand manner like this. People use languages they are more comfortable with, or that they feel are good for their use case. Tagged unions being able to represent ASTs easily is a big reason people use functional language for lang dev. There was also a survey in the /r/PL server recently and Rust was the 3rd most popular implementation language.

You've also already given Rust praise in the article so it seems a bit strange. 

Instead I recommend sticking to positives, what languages should people use?

Reply:
It's valid advice, but I rephrased it to be a little less aggressive.